File: Array/Gas Station/By iterative.cpp
// Method - Iterative(BruteForce)
// Time Complexity T - O(n^2), nested loop runs for n^2
// Space Complexity S - O(1), Bruteforce runs in constant sapce


class Solution {
public:
    int canCompleteCircuit(vector<int>& gas, vector<int>& cost) {
        
        // ensure the gas and cost are not null or empty
        // if not then return -1
        if(gas.size() == 0 || cost.size() == 0) {
            return -1;
        }
        
        // getting the no. of the gas station
        int n = gas.size();
        
        // outter loop runs for 0 to n-1
        for (int i = 0; i < n; i++) {
            
            // the remaining gas in the task
            // starting it is "0"
            int remainGas = 0;
            bool isSuccess = true;
            
            // inner loop runs for i to n+i
            for (int j = i; j < n + i; j++) {
                
                int station = (j+n) % n;
                
                remainGas = remainGas - cost[station] + gas[station] ;
                
                // if gas in tank becomes less than "0", 
                // then break inner loop and increment the outer loop
                if (remainGas < 0) {
                    isSuccess = false;
                    break;
                }
                
            }
            // if success
            // then return the "i"
            if (isSuccess == true) {
                return i;
            }
        }
        
        // if no succes not turn to 0 
        // then return -1
        return -1;
    }
};

File: Array/First Missing Positive Number/By iterative.java
// Method - Iterative(Optimised)
// Time Complexity T - O(n)
// Space Complexity S - O(1)

class Solution {
    public int firstMissingPositive(int[] nums) {
        
        // ensure the nums is not empty
        if (nums.length == 0){
            return 1;
        }    
        
        // loop runs for 0 to n-1 times
        for(int i = 0; i < nums.length;){
            
            // check nums[i] is not greater than size of num
            // or nums[i] values is not negative
            // if so increment "i"
            if(nums[i] > nums.length || nums[i] <= 0 ){
                i++;
            }
            
            // check nums[nums[i]-1] is equal to nums[i]
            else if(nums[nums[i]-1] == nums[i]){
                i++;
            }
            // else swap the numbers in correct order
            else{
                int temp = nums[nums[i]-1];
                nums[nums[i]-1] = nums[i];
                nums[i] = temp;
            }
        }
        
        int i=0;
        
        // loop rus for o - "n"
        for(i=0;i<nums.length; i++){
            // check the nums[i]= i+1
            // if not then return i+1
            if (nums[i] != i+1){
                return i+1;
            }
        }
        return nums[i-1]+1;
    }
}
